### PR TITLE
Reinstate deep JSON inspection of labeled metric tests

### PR DESCRIPTION
**Pull request by mdboom**
This brings back in some testing content from glean-ac in LabeledMetricTypeTests.kt.

Mainly, this checks pings against the schema, since getting that right for labeled metrics is a bit fiddly.

In order to do this, it makes `collectAndCheckPingSchema` from glean-ac work and updates the API to use `PingType` rather than ping names.

This had knockon effects in the FFI, and then the C FFI example, so that's been updated to run on Linux and also run in CI.  **I plan to go back and make that work on Mac as well again, once I've verified CI is working on that.**

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - `make test` runs without emitting any warnings
  - `make lint` runs without emitting any errors
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry or an explanation of why it does not need one
  - Any breaking changes to language binding APIs are noted explicitly

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -100,6 +100,18 @@ jobs:
     steps:
       - rust-tests:
           rust-version: "beta"
+  C tests:
+    docker:
+      - image: circleci/rust:latest
+    # We have to use a machine with more RAM for tests so we don't run out of memory.
```
Comment (badboy) [inline .circleci/config.yml]: Do we?

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt [hunk 2]
```diff
@@ -277,9 +278,12 @@ open class GleanInternalAPI internal constructor () {
...
 
-    fun collect(pingName: String) {
-        val s = LibGleanFFI.INSTANCE.glean_ping_collect(handle, pingName)!!
-        LibGleanFFI.INSTANCE.glean_str_free(s)
+    /**
+     * Collect a ping and return a string
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.NONE)
+    internal fun collect(ping: PingType): String? {
+        return LibGleanFFI.INSTANCE.glean_ping_collect(handle, ping.handle)?.getAndConsumeRustString()
```
Comment (badboy) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: that means we can close https://bugzilla.mozilla.org/show_bug.cgi?id=1552479

File: glean-core/ffi/examples/Makefile [hunk 3]
```diff
@@ -1,4 +1,4 @@
-objects = glean_app.c ../../../target/release/libglean_ffi.dylib
+objects = glean_app.c ../../../target/release/libglean_ffi.so
```
Comment (badboy) [inline glean-core/ffi/examples/Makefile]: Ah yeah, we probably need to properly detect what system we're on.

**Comment by badboy** [comment]
(Oh, this is stil in draft mode.)

File: .circleci/config.yml (hunk 1 above)
Comment (mdboom) [reply .circleci/config.yml]: Sorry -- copy-and-pasted from the Rust tests.  I doubt the C test is memory intensive...

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt (hunk 2 above)
Comment (mdboom) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: Yeah -- I initially thought we'd take it out, but it's super useful as a test-only function.  I say we close that bug when this PR lands.

File: .circleci/config.yml (hunk 1 above)
Comment (mdboom) [reply .circleci/config.yml]: Sorry -- copy-and-pasted from the Rust tests.  I doubt the C test is memory intensive...

File: glean-core/ffi/examples/Makefile (hunk 3 above)
Comment (mdboom) [reply glean-core/ffi/examples/Makefile]: I'm just handling this with the use of `$(wildcard )` now which isn't really detecting the system, but should be good enough for now...

Comment (badboy) [reply glean-core/ffi/examples/Makefile]: hm, probably still breaks on Windows, but good enough for now I'd say

Comment (mdboom) [reply glean-core/ffi/examples/Makefile]: Yeah, but it never worked on Windows ;)  Maybe we leave this to @Dexterp37 to decide whether to handle now (or later)...

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt [hunk 4]
```diff
@@ -277,9 +278,12 @@ open class GleanInternalAPI internal constructor () {
...
     }
 
-    fun collect(pingName: String) {
-        val s = LibGleanFFI.INSTANCE.glean_ping_collect(handle, pingName)!!
-        LibGleanFFI.INSTANCE.glean_str_free(s)
+    /**
+     * Collect a ping and return a string
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.NONE)
+    internal fun collect(ping: PingType): String? {
```
Comment (Dexterp37) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: nit: do we want to follow the same naming schema as other test only functions, e.g. `testCollect`?

Comment (mdboom) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt]: Yeah -- that's a good idea.

**Pull request merged**